services/mpa/mpa.proto

```diff
@@ -0,0 +1,120 @@
+/* Copyright (c) 2022 Snowflake Inc. All rights reserved.
+
+   Licensed under the Apache License, Version 2.0 (the
+   "License"); you may not use this file except in compliance
+   with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing,
+   software distributed under the License is distributed on an
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License.
+*/
+
+syntax = "proto3";
+
+option go_package = "github.com/Snowflake-Labs/sansshell/mpa";
+
+import "google/protobuf/any.proto";
+
+package Mpa;
+
+// The Mpa service definition
+service Mpa {
+  // Store a request that we'd like to try running in the future. Requests
+  // are stored in-memory and older requests may be cleared automatically.
+  // This call is idempotent - requests from the same user with the same
+  // contents will return the same id.
+  //
+  // The user for the request and any justification is implicitly passed in
+  // via inspecting the peer of the RPC or via gRPC metadata.
+  rpc Store(StoreRequest) returns (StoreResponse) {}
+  // Approve a previously stored request. A request can have multiple approvals
+  // and can be approved by anybody but the original user that stored the request.
+  //
+  // The user for the request is implicitly passed in via inspecting the
+  // peer of the RPC or via gRPC metadata.
+  rpc Approve(ApproveRequest) returns (ApproveResponse) {}
+  // Block until at least one approval has been granted. This is used
+  // as an optimization to avoid needing to poll for MPA approval.
+  rpc WaitForApproval(WaitForApprovalRequest) returns (WaitForApprovalResponse) {}
+  // List available requests.
+  rpc List(ListRequest) returns (ListResponse) {}
+  // Get a request and all approvals associated with it.
+  rpc Get(GetRequest) returns (GetResponse) {}
+  // Clear a stored request.
+  //
+  // This is typically unnecessary due to how requests are stored in memory
+  // and forgotten after a process restart.
+  rpc Clear(ClearRequest) returns (ClearResponse) {}
+}
+
+message Action {
+  // The user that created the request.
+  string user = 1;
+  // User-supplied information on why the request is being made.
+  string justification = 2;
+  // The GRPC method name, as '/Package.Service/Method'
+  string method = 3;
+  // The request protocol buffer.
+  google.protobuf.Any message = 4;
+}
+
+message Principal {
+  // The principal identifier (e.g. a username or service role)
+  string id = 1;
+  // Auxiliary groups associated with this principal.
+  repeated string groups = 2;
+}
+
+
+message StoreRequest {
```

Comment:
It might be helpful for StoreRequest (and StoreResponse) to have a TTL.  Eventually you'll want to expire requests that weren't consumed, or it'll OOM sansshell-server in the limit conditions.  A client default of 72hrs or something probably puts an upper bound on the memory usage, without unnecessarily hampering typical usage.  WDYT?

Reply:
Is there also a plan to cap the number of in-flight MPA requests that sansshell-server is willing to queue?  That would be another defense that might make the TTL unnecessary.

Reply:
I have a "Requests are stored in-memory and older requests may be cleared automatically" note in the RPC method comment that' should cover this.

I did "evict oldest request if there's >1000" in my prototype.  It's probably worth adding an "invalidate after X hours" so that old approvals don't cause confusion later on.